keps/2937-resource-transformer/README.md

```diff
@@ -0,0 +1,619 @@
+      Name:      cpu
+      Total:     1
+      Borrowed:  0
+      Name:      example.com/accelerator-memory
+      Total:     20G
+      Borrowed:  0
+      Name:      memory
+      Total:     97656250Ki
+  Flavors Usage:
+    Name:  default-flavor
+    Resources:
+      Borrowed:         0
+      Name:             cpu
+      Total:            1
+      Borrowed:         0
+      Name:             example.com/accelerator-memory
+      Total:            20G
+      Borrowed:         0
+      Name:             memory
+      Total:            97656250Ki
+  Pending Workloads:    0
+  Reserving Workloads:  1
+```
+
+After enough Workloads are admitted that there is queuing,
+then the detailed status message in the `QuotaReserved` condition
+of the Workload will indicate the resource(s) with insufficient quota, but the
+actual transformed resource request is not observable.
+```yaml
+Spec:
+  Active:  true
+  Pod Sets:
+    Count:  1
+    Name:   main
+    Template:
+      Metadata:
+      Spec:
+        Containers:
+          ...
+          Resources:
+            Limits:
+              nvidia.com/mig-1g.5gb:   2
+              nvidia.com/mig-2g.10gb:  1
+            Requests:
+              Cpu:                     1
+              Memory:                  100G
+              nvidia.com/mig-1g.5gb:   2
+              nvidia.com/mig-2g.10gb:  1
+...
+Status:
+  Conditions:
+    Last Transition Time:  2024-09-19T14:59:01Z
+    Message:               couldn't assign flavors to pod set main: insufficient unused quota for example.com/accelerator-memory in flavor default-flavor, 20G more needed
+    Observed Generation:   1
+    Reason:                Pending
+    Status:                False
+    Type:                  QuotaReserved
+```
+
+We could improve the observability by extending the Workload Status with a ResourceRequests
+field that would contain a copy of the `TotalRequests` field of the controller-internal `workload.Info`
+struct.  If we did this, we would instead see something like the below for a non-admitted Workload:
+```yaml
+Spec:
+  Active:  true
+  Pod Sets:
+    Count:  1
+    Name:   main
+    Template:
+      Metadata:
+      Spec:
+        Containers:
+          ...
+          Resources:
+            Limits:
+              nvidia.com/mig-1g.5gb:   2
+              nvidia.com/mig-2g.10gb:  1
+            Requests:
+              Cpu:                     1
+              Memory:                  100G
+              nvidia.com/mig-1g.5gb:   2
+              nvidia.com/mig-2g.10gb:  1
+...
+Status:
+  Conditions:
+    Last Transition Time:  2024-09-19T14:59:01Z
+    Message:               couldn't assign flavors to pod set main: insufficient unused quota for example.com/accelerator-memory in flavor default-flavor, 20G more needed
+    Observed Generation:   1
+    Reason:                Pending
+    Status:                False
+    Type:                  QuotaReserved
+  ResourceRequests:
+    Name: main
+    Resource Request:
```

Comment:
I'm happy to adjust both the KEP (in a follow up PR) and the prototype implementation so we end up in a good place.

Reply:
> It crossed my mind that there might be duplicated information within the admission field.

Yes, that's true, but considering the case of partial admission, dynamic reclaiming and dynamically sized Jobs (not yet implemented), a dedicated transformed resource request field might be worth it since the admission field has only admitted resource requests rather than a whole of input resource requests.

Although, adding the transformed resource information to admission field might be better, I guess that we can consider it as a separate enhancement.

> But if we go with the name TransformedPodSetResources, then we would only list the transformed resources, and not the original ones.

Good point. That sounds reasonable. We should hold only modified information in the status field. The immutable field should not be added to the status field.

Reply:
Let's continue the discussion in the implementation PR.
We can refine the design and API in the implementation PR.